src/main/java/org/assertj/core/api/RecursiveIterableComparisonAssert.java

```diff
@@ -0,0 +1,285 @@
+package org.assertj.core.api;
+
+import org.assertj.core.api.recursive.comparison.RecursiveComparisonConfiguration;
+import org.assertj.core.util.Arrays;
+import org.assertj.core.util.introspection.IntrospectionError;
+
+import static org.assertj.core.util.IterableUtil.sizeOf;
+
+public class RecursiveIterableComparisonAssert<SELF extends RecursiveIterableComparisonAssert<SELF>> extends RecursiveComparisonAssert<SELF> {
+
+  private RecursiveComparisonConfiguration recursiveComparisonConfiguration;
+  private AbstractIterableAssert iterableAssert;
+
+  public RecursiveIterableComparisonAssert(Object actual, RecursiveComparisonConfiguration recursiveComparisonConfiguration, AbstractIterableAssert iterableAssert) {
+    super(actual, recursiveComparisonConfiguration);
+    this.iterableAssert = iterableAssert;
+    this.recursiveComparisonConfiguration = recursiveComparisonConfiguration;
+  }
+
+  /**
+   * By default, any type of iterables can be compared with each other. This may cause unexpected behavior when one of them is ordered and other is not.
+   * For example, a List can be compared to a Set and even though they might have exact same elements, a Set will not preserve insertion order.
+   * So the assertion will fail. To avoid these kind of surprises, comparison ordered can be ignored with calling {@link #ignoringActualIterableOrder ignoringActualIterableOrder}.
+   * <p>
+   * Example:
+   * <pre><code class='java'>public class Person {
+   *     String name;
+   *     boolean hasPhd;
+   * }
+   *
+   * Person sheldon = new People("Sheldon Cooper", true);
+   * Person howard = new People("Howard Wolowitz", false);
+   * Person penny = new People("Penny", false);
+   *
+   * List<Person> peopleList = Arrays.asList(sheldon, howard, penny);
+   * Set<Person> peopleSet = new HashSet<>(peopleList);
+   *
+   * // assertion fails even though iterables contain the exact same elements because order is different.
+   * assertThat(peopleList).usingRecursiveComparison()
+   *                       .isEqualTo(peopleSet);
+   *
+   * // assertion succeeds because both iterables contain exact same items and comparison order is ignored.
+   * assertThat(peopleList).usingRecursiveComparison()
+   *                       .ignoringActualIterableOrder()
+   *                       .isEqualTo(peopleArray);</code></pre>
+   *
+   * @return this {@link RecursiveIterableComparisonAssert} to chain other methods.
+   */
+  public SELF ignoringActualIterableOrder() {
```

Comment:
I would rename it to `ignoringIterableOrder`. `ignoringElementsOrder` or `ignoringOrder` as we ignore as much actual order as we ignore expected order.

Also I expect this to also ignore order for any element fields that are iterable too (the implementation might already do that, I haven't checked) which should be documented in the javadoc.

Reply:
The RecursiveComparisonAssert API already has `ignoringCollectionOrder` methods, what use case `ignoringActualIterableOrder` covers that these methods don't.

Reply:
From my understanding, `RecursiveComparisonAssert#ignoringCollectionOrder` works when the **field** of an element is an iterable, not the element itself, which is the case for `RecursiveIterableComparisonAssert`. For example:

```
Person p1 = new Person("harun", 1);
Person p2 = new Person("harun", 2);
Person p3 = new Person("harun", 3);
List<Person> list1 = Arrays.asList(p1, p2, p3);
List<Person> list2 = Arrays.asList(p2, p3, p1);
assertThat(list1).usingRecursiveComparison().ignoringCollectionOrder().isEqualTo(list2);
```

This assertion fails despite the usage of `ignoringCollectionOrder`, since `ignoringCollectionOrder `only applies to the **fields** under comparison and not the actual collection as a whole. Based on my understanding, I thought that I should provide a method that would actually ignore the order of actual/expected iterables.

Reply:
The best way forward is to make `ignoringCollectionOrder` effective on actual iterables, the rationale being that what applies for fields should also apply for top level objects.